clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1160,15 +1160,35 @@ static target getAccessTarget(QualType FieldTy,
 }
 
 bool SemaSYCL::isFreeFunction(const FunctionDecl *FD) {
-  for (auto *IRAttr : FD->specific_attrs<SYCLAddIRAttributesFunctionAttr>()) {
-    SmallVector<std::pair<std::string, std::string>, 4> NameValuePairs =
-        IRAttr->getAttributeNameValuePairs(getASTContext());
-    for (const auto &NameValuePair : NameValuePairs) {
-      if (NameValuePair.first == "sycl-nd-range-kernel" ||
-          NameValuePair.first == "sycl-single-task-kernel") {
-        return true;
+  llvm::SmallVector<clang::FunctionDecl *, 4> Redecls;
+  for (clang::FunctionDecl *Redecl : FD->redecls()) {
+    Redecls.push_back(Redecl);
+  }
+  bool FirstDecl = true;
+  clang::SourceLocation Loc = FD->getLocation();
+  while (!Redecls.empty()) {
```

Comment:
AFAIK a FunctionDecl should have a member `isFirstDecl`. Can we just loop over `FD->redecls()`, and issue an error in case an attribute is applied to a decl returning false from `isFirstDecl`?